File: python/download_tiles.py
"""
Download MapTiler satellite tiles for a fixed bounding box and save them
as static assets under  webapp/public/tiles/satellite/{z}/{x}/{y}.jpg

Run once (or whenever you want to refresh imagery):
    python python/download_tiles.py
"""

import math
import os
import time
import urllib.request

# ── Configuration ──────────────────────────────────────────────────────────────
API_KEY = "ggrNEhS6ufFFsGzNwRzy"

# Alversjö centre
CENTER_LON = 14.923
CENTER_LAT = 57.620

# How much padding (in degrees) around the centre to download.
# ~0.01° ≈ 1 km at this latitude — gives a comfortable margin.
PAD_LON = 0.012
PAD_LAT = 0.008

ZOOM_MIN = 13
ZOOM_MAX = 17

TILE_URL = "https://api.maptiler.com/tiles/satellite-v2/{z}/{x}/{y}.jpg?key={key}"
# Script lives in python/, so go up one level to project root, then into webapp/public
OUT_DIR = os.path.join(os.path.dirname(__file__), "..", "webapp", "public", "tiles", "satellite")

# ── Helpers ────────────────────────────────────────────────────────────────────

def lon_to_tile_x(lon: float, zoom: int) -> int:
    """Convert longitude to tile X index."""
    return int(math.floor((lon + 180) / 360 * (1 << zoom)))


def lat_to_tile_y(lat: float, zoom: int) -> int:
    """Convert latitude to tile Y index (Web‑Mercator / slippy‑map)."""
    lat_rad = math.radians(lat)
    n = 1 << zoom
    return int(math.floor((1 - math.log(math.tan(lat_rad) + 1 / math.cos(lat_rad)) / math.pi) / 2 * n))


def download_tile(z: int, x: int, y: int) -> bool:
    """Download a single tile and save it.  Returns True on success."""
    url = TILE_URL.format(z=z, x=x, y=y, key=API_KEY)
    dest = os.path.join(OUT_DIR, str(z), str(x), f"{y}.jpg")
    if os.path.exists(dest):
        return True  # already cached

    os.makedirs(os.path.dirname(dest), exist_ok=True)
    try:
        urllib.request.urlretrieve(url, dest)
        return True
    except Exception as exc:
        print(f"  ✗ {z}/{x}/{y}  –  {exc}")
        return False


# ── Main ───────────────────────────────────────────────────────────────────────

def main():
    bbox_west = CENTER_LON - PAD_LON
    bbox_east = CENTER_LON + PAD_LON
    bbox_south = CENTER_LAT - PAD_LAT
    bbox_north = CENTER_LAT + PAD_LAT

    total = 0
    downloaded = 0

    for z in range(ZOOM_MIN, ZOOM_MAX + 1):
        x_min = lon_to_tile_x(bbox_west, z)
        x_max = lon_to_tile_x(bbox_east, z)
        y_min = lat_to_tile_y(bbox_north, z)  # note: y is inverted
        y_max = lat_to_tile_y(bbox_south, z)

        count = (x_max - x_min + 1) * (y_max - y_min + 1)
        print(f"Zoom {z}: tiles x=[{x_min}..{x_max}]  y=[{y_min}..{y_max}]  ({count} tiles)")

        for x in range(x_min, x_max + 1):
            for y in range(y_min, y_max + 1):
                total += 1
                if download_tile(z, x, y):
                    downloaded += 1
                time.sleep(0.05)  # gentle rate‑limit

    print(f"\nDone — {downloaded}/{total} tiles saved to {os.path.abspath(OUT_DIR)}")


if __name__ == "__main__":
    main()

File: webapp/scripts/download_tiles.mjs
/**
 * Download MapTiler satellite tiles for a fixed bounding box and save them
 * as static assets under  public/tiles/satellite/{z}/{x}/{y}.jpg
 *
 * Run once (or whenever you want to refresh imagery):
 *     node scripts/download_tiles.mjs
 */

import fs from 'fs';
import https from 'https';
import path from 'path';
import { fileURLToPath } from 'url';

// ── Configuration ─────────────────────────────────────────────────────────────
const API_KEY = 'ggrNEhS6ufFFsGzNwRzy';

// Alversjö centre
const CENTER_LON = 14.923;
const CENTER_LAT = 57.620;

// Padding in degrees (~0.012° ≈ 1 km at this latitude)
const PAD_LON = 0.012;
const PAD_LAT = 0.008;

const ZOOM_MIN = 13;
const ZOOM_MAX = 17;

const TILE_URL = (z, x, y) =>
  `https://api.maptiler.com/tiles/satellite-v2/${z}/${x}/${y}.jpg?key=${API_KEY}`;

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const OUT_DIR = path.resolve(__dirname, '..', 'public', 'tiles', 'satellite');

// ── Helpers ───────────────────────────────────────────────────────────────────

function lonToTileX(lon, zoom) {
  return Math.floor(((lon + 180) / 360) * (1 << zoom));
}

function latToTileY(lat, zoom) {
  const latRad = (lat * Math.PI) / 180;
  const n = 1 << zoom;
  return Math.floor(
    ((1 - Math.log(Math.tan(latRad) + 1 / Math.cos(latRad)) / Math.PI) / 2) * n
  );
}

function downloadFile(url, dest) {
  return new Promise((resolve, reject) => {
    const dir = path.dirname(dest);
    fs.mkdirSync(dir, { recursive: true });

    const file = fs.createWriteStream(dest);
    https
      .get(url, (res) => {
        if (res.statusCode !== 200) {
          file.close();
          fs.unlinkSync(dest);
          reject(new Error(`HTTP ${res.statusCode} for ${url}`));
          return;
        }
        res.pipe(file);
        file.on('finish', () => {
          file.close();
          resolve();
        });
      })
      .on('error', (err) => {
        file.close();
        if (fs.existsSync(dest)) fs.unlinkSync(dest);
        reject(err);
      });
  });
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  const bboxWest = CENTER_LON - PAD_LON;
  const bboxEast = CENTER_LON + PAD_LON;
  const bboxSouth = CENTER_LAT - PAD_LAT;
  const bboxNorth = CENTER_LAT + PAD_LAT;

  let total = 0;
  let downloaded = 0;

  for (let z = ZOOM_MIN; z <= ZOOM_MAX; z++) {
    const xMin = lonToTileX(bboxWest, z);
    const xMax = lonToTileX(bboxEast, z);
    const yMin = latToTileY(bboxNorth, z); // y is inverted
    const yMax = latToTileY(bboxSouth, z);

    const count = (xMax - xMin + 1) * (yMax - yMin + 1);
    console.log(`Zoom ${z}: tiles x=[${xMin}..${xMax}]  y=[${yMin}..${yMax}]  (${count} tiles)`);

    for (let x = xMin; x <= xMax; x++) {
      for (let y = yMin; y <= yMax; y++) {
        total++;
        const dest = path.join(OUT_DIR, String(z), String(x), `${y}.jpg`);
        if (fs.existsSync(dest)) {
          downloaded++;
          continue; // already cached
        }
        try {
          await downloadFile(TILE_URL(z, x, y), dest);
          downloaded++;
          process.stdout.write('.');
          await sleep(50); // gentle rate-limit
        } catch (err) {
          console.error(`\n  ✗ ${z}/${x}/${y}  –  ${err.message}`);
        }
      }
    }
    console.log();
  }

  console.log(`\nDone — ${downloaded}/${total} tiles saved to ${OUT_DIR}`);
}

main();
